File: src/window/linux.c
#include "../window.h"

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/XKBlib.h>
#include <X11/keysym.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <linux/joystick.h>
#include <string.h>

static Display *g_display;
static Window g_window;
static Atom g_atom_protocols;
static Atom g_atom_delete_window;
static int g_fd = -1;
static int32_t g_input_states[NUM_OF_KEYCODES];

static SsKeycode convert_keycode_xcb_to_ss(KeySym key_sym) {
    switch (key_sym) {
        case XK_a: return KEYCODE_A;
        case XK_b: return KEYCODE_B;
        case XK_c: return KEYCODE_C;
        case XK_d: return KEYCODE_D;
        case XK_e: return KEYCODE_E;
        case XK_f: return KEYCODE_F;
        case XK_g: return KEYCODE_G;
        case XK_h: return KEYCODE_H;
        case XK_i: return KEYCODE_I;
        case XK_j: return KEYCODE_J;
        case XK_k: return KEYCODE_K;
        case XK_l: return KEYCODE_L;
        case XK_m: return KEYCODE_M;
        case XK_n: return KEYCODE_N;
        case XK_o: return KEYCODE_O;
        case XK_p: return KEYCODE_P;
        case XK_q: return KEYCODE_Q;
        case XK_r: return KEYCODE_R;
        case XK_s: return KEYCODE_S;
        case XK_t: return KEYCODE_T;
        case XK_u: return KEYCODE_U;
        case XK_v: return KEYCODE_V;
        case XK_w: return KEYCODE_W;
        case XK_x: return KEYCODE_X;
        case XK_y: return KEYCODE_Y;
        case XK_z: return KEYCODE_Z;
        case XK_0: return KEYCODE_0;
        case XK_1: return KEYCODE_1;
        case XK_2: return KEYCODE_2;
        case XK_3: return KEYCODE_3;
        case XK_4: return KEYCODE_4;
        case XK_5: return KEYCODE_5;
        case XK_6: return KEYCODE_6;
        case XK_7: return KEYCODE_7;
        case XK_8: return KEYCODE_8;
        case XK_9: return KEYCODE_9;
        case XK_Up: return KEYCODE_UP;
        case XK_Left: return KEYCODE_LEFT;
        case XK_Down: return KEYCODE_DOWN;
        case XK_Right: return KEYCODE_RIGHT;
        case XK_Return: return KEYCODE_ENTER;
        case XK_space: return KEYCODE_SPACE;
        case XK_Shift_L:
        case XK_Shift_R: return KEYCODE_SHIFT;
        case XK_Tab: return KEYCODE_TAB;
        case XK_Control_L:
        case XK_Control_R: return KEYCODE_CONTROL;
        case XK_Escape: return KEYCODE_ESCAPE;
        default: return NO_KEYCODE;
    }
}

void create_window_param(WindowParam *out) {
    out->xlib.display = g_display;
    out->xlib.window = g_window;
}

warn_t create_window(const char *title, uint32_t width, uint32_t height) {
    ss_info("initializing xlib g_window ...");
    ss_indent_logger();

    g_display = XOpenDisplay(NULL);
    if (g_display == NULL)
        ss_error("failed to open g_display.");
    // TODO: error handling
    g_window = XCreateSimpleWindow(g_display, RootWindow(g_display, 0), 0, 0, width, height, 0, 0, WhitePixel(g_display, 0));

    XStoreName(g_display, g_window, title);

    XSizeHints size_hints;
    memset(&size_hints, 0, sizeof(XSizeHints));
    size_hints.flags = PMinSize | PMaxSize;
    size_hints.max_width = (int32_t)width;
    size_hints.min_width = (int32_t)width;
    size_hints.max_height = (int32_t)height;
    size_hints.min_height = (int32_t)height;
    XSetWMNormalHints(g_display, g_window, &size_hints);

    g_atom_protocols = XInternAtom(g_display, "WM_PROTOCOLS", False);
    g_atom_delete_window = XInternAtom(g_display, "WM_DELETE_WINDOW", False);
    XSetWMProtocols(g_display, g_window, &g_atom_delete_window, 1);


    memset(g_input_states, 0, sizeof(int32_t) * NUM_OF_KEYCODES);
    XSelectInput(g_display, g_window, KeyPressMask | KeyReleaseMask);
    if ((g_fd = open("/dev/input/js0", O_RDONLY | O_NONBLOCK)) < 0)
        ss_info("joystick not detected.");
    else
        ss_info("joystick detected.");

    XMapWindow(g_display, g_window);
    XFlush(g_display);

    ss_dedent_logger();
    ss_info("xlib window initialization succeeded.");
    return SS_SUCCESS;
}

int32_t do_window_events(void) {
    const size_t changes_cnt = NUM_OF_KB_KEYS + NUM_OF_JS_BUTTONS;
    char input_state_changes[changes_cnt];
    memset(input_state_changes, 0, sizeof(char) * changes_cnt);

    // do events and get key input state changes
    XEvent event;
    while (XPending(g_display)) {
        XNextEvent(g_display, &event);
        switch (event.type) {
            case ClientMessage:
                if (event.xclient.message_type == g_atom_protocols && event.xclient.data.l[0] == g_atom_delete_window)
                    return 1;
            case KeyPress: {
                const KeySym key_sym = XkbKeycodeToKeysym(g_display, event.xkey.keycode, 0, 0);
                const SsKeycode ss_keycode = convert_keycode_xcb_to_ss(key_sym);
                input_state_changes[ss_keycode] = 1;
                break;
            }
            case KeyRelease: {
                const KeySym key_sym = XkbKeycodeToKeysym(g_display, event.xkey.keycode, 0, 0);
                const SsKeycode ss_keycode = convert_keycode_xcb_to_ss(key_sym);
                input_state_changes[ss_keycode] = 2;
                break;
            }
        }
    }

    // monitor joystick
    if (g_fd >= 0) {
        struct js_event jevent;
        const size_t js_event_size = sizeof(struct js_event);
        ssize_t len = read(g_fd, &jevent, js_event_size);
        if (len == js_event_size) {
            if ((jevent.type & JS_EVENT_BUTTON) && jevent.number < NUM_OF_JS_BUTTONS)
                input_state_changes[KEYCODE_JS_BUTTON_0 + jevent.number] = jevent.value == 1 ? 1 : 2;
            else if ((jevent.type & JS_EVENT_AXIS) && jevent.number < NUM_OF_JS_AXES)
                // NOTE: jevent.value is not 0/1.
                // NOTE: while the joystick is tilted, the event will continue to occur.
                g_input_states[KEYCODE_JS_AXIS_0 + jevent.number] = jevent.value;
        }
    }

    // update input states
    for (int i = 1; i < changes_cnt; ++i) {
        if (input_state_changes[i] == 1 && g_input_states[i] < 1)
            g_input_states[i] = 1;
        else if (input_state_changes[i] == 2)
            g_input_states[i] = -1;
        else if (g_input_states[i] > 0)
            g_input_states[i] += 1;
        else if (g_input_states[i] < 0)
            g_input_states[i] = 0;
    }

    // finish
    return 0;
}

void terminate_window(void) {}

int32_t get_input_state(SsKeycode keycode) {
    return g_input_states[keycode];
}

File: src/window/linux.rs
#![allow(non_upper_case_globals)]

use crate::log::*;
use crate::tpl::*;

use super::Keycode;

use std::collections::HashMap;
use std::ffi::c_void;
use std::mem::size_of;
use std::os::raw::{c_char, c_int, c_long};
use std::ptr::null;

const WM_PROTOCOLS: *const c_char = [
    'W' as c_char,
    'M' as c_char,
    '_' as c_char,
    'P' as c_char,
    'R' as c_char,
    'O' as c_char,
    'T' as c_char,
    'O' as c_char,
    'C' as c_char,
    'O' as c_char,
    'L' as c_char,
    'S' as c_char,
    '\0' as c_char,
]
.as_ptr();
const WM_DELETE_WINDOW: *const c_char = [
    'W' as c_char,
    'M' as c_char,
    '_' as c_char,
    'D' as c_char,
    'E' as c_char,
    'L' as c_char,
    'E' as c_char,
    'T' as c_char,
    'E' as c_char,
    '_' as c_char,
    'W' as c_char,
    'I' as c_char,
    'N' as c_char,
    'D' as c_char,
    'O' as c_char,
    'W' as c_char,
    '\0' as c_char,
]
.as_ptr();

const DEV_INPUT_JS0: *const c_char = [
    '/' as c_char,
    'd' as c_char,
    'e' as c_char,
    'v' as c_char,
    '/' as c_char,
    'i' as c_char,
    'n' as c_char,
    'p' as c_char,
    'u' as c_char,
    't' as c_char,
    '/' as c_char,
    'j' as c_char,
    's' as c_char,
    '0' as c_char,
    '\0' as c_char,
]
.as_ptr();

const POV_THRESHOLD: __s16 = 30000;

pub struct WindowApp {
    pub(crate) display: *mut Display,
    pub(crate) window: Window,
    atom_protocols: Atom,
    atom_delete_window: Atom,
    input_states: HashMap<Keycode, i32>,
    fd: c_int,
}

impl WindowApp {
    /// A constructor.
    /// It creates and shows a window.
    pub fn new(title: &str, width: u32, height: u32) -> Self {
        // create a window
        let display = unsafe { XOpenDisplay(null()) };
        if display.is_null() {
            ss_error("failed to create a display in Xlib.");
        }
        let window = unsafe {
            XCreateSimpleWindow(
                display,
                XDefaultRootWindow(display),
                0,
                0,
                width,
                height,
                0,
                0,
                0,
            )
        };

        // change the window title
        let mut c_title = title.bytes().map(|c| c as c_char).collect::<Vec<c_char>>();
        c_title.push(0);
        let c_title = c_title.as_ptr();
        unsafe { XStoreName(display, window, c_title) };

        // make the window unresizable
        let mut size_hints = XSizeHints {
            flags: (PMinSize | PMaxSize) as i64,
            x: 0,
            y: 0,
            width: 0,
            height: 0,
            min_width: width as i32,
            min_height: height as i32,
            max_width: width as i32,
            max_height: height as i32,
            width_inc: 0,
            height_inc: 0,
            min_aspect: XSizeHints__bindgen_ty_1 { x: 0, y: 0 },
            max_aspect: XSizeHints__bindgen_ty_1 { x: 0, y: 0 },
            base_width: 0,
            base_height: 0,
            win_gravity: 0,
        };
        unsafe { XSetWMNormalHints(display, window, &mut size_hints) };

        // set the window deletion event
        let atom_protocols = unsafe { XInternAtom(display, WM_PROTOCOLS, False as c_int) };
        let mut atom_delete_window =
            unsafe { XInternAtom(display, WM_DELETE_WINDOW, False as c_int) };
        unsafe { XSetWMProtocols(display, window, &mut atom_delete_window, 1) };

        // set the window input event
        unsafe { XSelectInput(display, window, (KeyPressMask | KeyReleaseMask) as c_long) };

        // open /dev/input/js0
        let fd = unsafe { open(DEV_INPUT_JS0, O_RDONLY as c_int | O_NONBLOCK as c_int) };

        // show the window
        unsafe { XMapWindow(display, window) };
        unsafe { XFlush(display) };

        // finish
        Self {
            display,
            window,
            atom_protocols,
            atom_delete_window,
            input_states: HashMap::new(),
            fd,
        }
    }

    /// A method to do all queued window events and check if the window is deleted.
    /// If the window is deleted it returns `false`, otherwise `true`.
    pub fn do_events(&mut self) -> bool {
        // do all window events
        while unsafe { XPending(self.display) } != 0 {
            // get a window event
            let mut event = _XEvent { type_: 0 };
            unsafe { XNextEvent(self.display, &mut event) };

            // get the event type
            let type_ = unsafe { event.type_ as u32 };

            // when it's a window deletion event, tell invoker to end the process
            match type_ {
                ClientMessage => {
                    let flag_type = unsafe { event.xclient.message_type == self.atom_protocols };
                    let flag_data =
                        unsafe { event.xclient.data.l[0] == self.atom_delete_window as c_long };
                    if flag_type && flag_data {
                        return false;
                    }
                }
                KeyPress => {
                    let key_sym = unsafe {
                        XkbKeycodeToKeysym(self.display, event.xkey.keycode as KeyCode, 0, 0)
                    };
                    if let Some(kc) = convert_x_to_ss(key_sym) {
                        let v = *self.input_states.entry(kc.clone()).or_insert_with(|| 0);
                        self.input_states.insert(kc, v);
                    }
                }
                KeyRelease => {
                    let key_sym = unsafe {
                        XkbKeycodeToKeysym(self.display, event.xkey.keycode as KeyCode, 0, 0)
                    };
                    if let Some(kc) = convert_x_to_ss(key_sym) {
                        self.input_states.remove(&kc);
                    }
                }
                _ => (),
            }
        }

        // update js0
        if self.fd >= 0 {
            let mut jevent = js_event {
                time: 0,
                value: 0,
                type_: 0,
                number: 0,
            };
            let js_event_size = size_of::<js_event>();
            if unsafe {
                crate::tpl::read(self.fd, &mut jevent as *mut _ as *mut c_void, js_event_size)
                    == js_event_size as isize
            } {
                if jevent.type_ as u32 & JS_EVENT_BUTTON > 0 {
                    if let Some(kc) = convert_js_to_ss(jevent.number) {
                        if jevent.value == 1 {
                            let v = *self.input_states.entry(kc.clone()).or_insert_with(|| 0);
                            self.input_states.insert(kc, v);
                        } else {
                            self.input_states.remove(&kc);
                        }
                    }
                }
                if jevent.type_ as u32 & JS_EVENT_AXIS > 0 {
                    // pov
                    if jevent.number == 6 {
                        self.update_axis(
                            Keycode::JsButtonRight,
                            Keycode::JsButtonLeft,
                            jevent.value,
                        );
                    } else if jevent.number == 7 {
                        self.update_axis(Keycode::JsButtonDown, Keycode::JsButtonUp, jevent.value);
                    }
                    // axis
                    else if jevent.number == 0 {
                        self.input_states
                            .insert(Keycode::JsAxisLX, jevent.value as i32);
                    } else if jevent.number == 1 {
                        self.input_states
                            .insert(Keycode::JsAxisLY, jevent.value as i32);
                    } else if jevent.number == 3 {
                        self.input_states
                            .insert(Keycode::JsAxisRX, jevent.value as i32);
                    } else if jevent.number == 4 {
                        self.input_states
                            .insert(Keycode::JsAxisRY, jevent.value as i32);
                    }
                }
            }
        }

        // update input states
        self.input_states.iter_mut().for_each(|(_, n)| *n += 1);
        true
    }

    /// A method to get a key or game pad button input state.
    /// It returns how many frames the key or button pressed.
    /// If you try to get game pad axis input state, it returns 0.
    pub fn get_input(&self, kc: Keycode) -> u32 {
        match kc {
            Keycode::JsAxisLX | Keycode::JsAxisLY | Keycode::JsAxisRX | Keycode::JsAxisRY => 0,
            _ => *self.input_states.get(&kc).unwrap_or(&0) as u32,
        }
    }

    /// A method to get a game pad axis input state.
    /// It returns the tangent of axis as `[-1.0, 1.0]`.
    /// If you try to get any input state except the axis state, it returns 0.
    pub fn get_axis_input(&self, kc: Keycode) -> f32 {
        match kc {
            Keycode::JsAxisLX | Keycode::JsAxisLY | Keycode::JsAxisRX | Keycode::JsAxisRY => {
                let state = *self.input_states.get(&kc).unwrap_or(&0);
                if state > 0 {
                    state as f32 / i16::MAX as f32
                } else {
                    state as f32 / i16::MIN as f32 * -1.0
                }
            }
            _ => 0.0,
        }
    }

    pub fn terminate(self) {}

    fn update_axis(&mut self, kc_greater: Keycode, kc_less: Keycode, value: __s16) {
        if value > POV_THRESHOLD {
            let v = *self
                .input_states
                .entry(kc_greater.clone())
                .or_insert_with(|| 0);
            self.input_states.insert(kc_greater, v);
        } else if value < -POV_THRESHOLD {
            let v = *self
                .input_states
                .entry(kc_less.clone())
                .or_insert_with(|| 0);
            self.input_states.insert(kc_less, v);
        } else {
            self.input_states.remove(&kc_greater);
            self.input_states.remove(&kc_less);
        }
    }
}

fn convert_x_to_ss(key_sym: KeySym) -> Option<Keycode> {
    match key_sym as u32 {
        XK_a => Some(Keycode::KeyA),
        XK_b => Some(Keycode::KeyB),
        XK_c => Some(Keycode::KeyC),
        XK_d => Some(Keycode::KeyD),
        XK_e => Some(Keycode::KeyE),
        XK_f => Some(Keycode::KeyF),
        XK_g => Some(Keycode::KeyG),
        XK_h => Some(Keycode::KeyH),
        XK_i => Some(Keycode::KeyI),
        XK_j => Some(Keycode::KeyJ),
        XK_k => Some(Keycode::KeyK),
        XK_l => Some(Keycode::KeyL),
        XK_m => Some(Keycode::KeyM),
        XK_n => Some(Keycode::KeyN),
        XK_o => Some(Keycode::KeyO),
        XK_p => Some(Keycode::KeyP),
        XK_q => Some(Keycode::KeyQ),
        XK_r => Some(Keycode::KeyR),
        XK_s => Some(Keycode::KeyS),
        XK_t => Some(Keycode::KeyT),
        XK_u => Some(Keycode::KeyU),
        XK_v => Some(Keycode::KeyV),
        XK_w => Some(Keycode::KeyW),
        XK_x => Some(Keycode::KeyX),
        XK_y => Some(Keycode::KeyY),
        XK_z => Some(Keycode::KeyZ),
        XK_0 => Some(Keycode::Key0),
        XK_1 => Some(Keycode::Key1),
        XK_2 => Some(Keycode::Key2),
        XK_3 => Some(Keycode::Key3),
        XK_4 => Some(Keycode::Key4),
        XK_5 => Some(Keycode::Key5),
        XK_6 => Some(Keycode::Key6),
        XK_7 => Some(Keycode::Key7),
        XK_8 => Some(Keycode::Key8),
        XK_9 => Some(Keycode::Key9),
        XK_Up => Some(Keycode::Up),
        XK_Left => Some(Keycode::Left),
        XK_Down => Some(Keycode::Down),
        XK_Right => Some(Keycode::Right),
        XK_Return => Some(Keycode::Enter),
        XK_space => Some(Keycode::Space),
        XK_Shift_L | XK_Shift_R => Some(Keycode::Shift),
        XK_Tab => Some(Keycode::Tab),
        XK_Control_L | XK_Control_R => Some(Keycode::Control),
        XK_Escape => Some(Keycode::Escape),
        _ => None,
    }
}

fn convert_js_to_ss(number: __u8) -> Option<Keycode> {
    match number {
        0 => Some(Keycode::JsButtonA),
        1 => Some(Keycode::JsButtonB),
        2 => Some(Keycode::JsButtonX),
        3 => Some(Keycode::JsButtonY),
        4 => Some(Keycode::JsButtonLShoulder),
        5 => Some(Keycode::JsButtonRShoulder),
        6 => Some(Keycode::JsButtonBack),
        7 => Some(Keycode::JsButtonStart),
        9 => Some(Keycode::JsButtonLThumb),
        10 => Some(Keycode::JsButtonRThumb),
        _ => None,
    }
}
